solo-io/gloo
# Vault renewal with refactor

# Description

Implement token renewal as demonstrated here: https://github.com/hashicorp/vault-examples/blob/main/examples/token-renewal/go/example.go

## API changes
Added `settings.secretOptions.sources[].vault.aws.leaseInterval` which gets used as the `Increment` argument when creating the `LifeTimewatcher` to monitor the token. If the value not defined, it will default to 0, which will be treated by Vault as the ttl of the secret.


## Code changes
Modified the `ClientAuth` interface by adding `StartRenewal`:
```
type ClientAuth interface {
	// vault.AuthMethod provides Login(ctx context.Context, client *Client) (*Secret, error)
	vault.AuthMethod
	// ManageTokenRenewal should be called after a successful login to start the renewal process
	// This method may have many different types of implementation, from just a noop to spinning up a separate go routine
	ManageTokenRenewal(ctx context.Context, client *vault.Client, secret *vault.Secret)
}

```
The renewal logic is in `projects/gloo/pkg/bootstrap/clients/vault/renewal.go`
New interfaces:
```
type TokenRenewer interface {
	StartRenewal(ctx context.Context, client *vault.Client, clientAuth ClientAuth, secret *vault.Secret) error
}
```

The `TokenRenewer` interface is used to renew tokens. We have implemented `VaultTokenRenewer` to handle the vault logic, and it can be replaced during testing. 
```
type VaultTokenRenewer struct {
	auth                     vault.AuthMethod
	leaseIncrement           int
	getWatcher               getWatcherFunc
	retryOnNonRenewableSleep int
}
```

The `TokenWatcher` interface is implemented by `vault.LifetimeWatcher`, and is broken out for testing purposes so we can send our own messages.
```
type TokenWatcher interface {
	DoneCh() <-chan error
	RenewCh() <-chan *vault.RenewOutput
        Start()
        Stop()
}
```
The TokenWatcher interface matches the shape of the LifetimeWatcher in our code and was implemented to let us switch out the implementation for testing.

We added the following metrics:

gloo_solo_io_vault_last_login_success - Time of last successful vault login
gloo_solo_io_vault_last_login_failure - Time of last failed vault login

gloo_solo_io_vault_last_renew_success - Time of last successful vault renewal
gloo_solo_io_vault_last_renew_failure - Time of last failed vault renewal - note, these are somewhat common and will happen when the token ttl has expired. This will be followed by a login attempt.

gloo_solo_io_vault_login_successes - Number of successful authentications of vault
gloo_solo_io_vault_login_failures - Number of failed authentications of vault

gloo_solo_io_vault_renew_successes - Number of successful vault renewals
gloo_solo_io_vault_renew_failures - Number of successful vault failures

## Docs changes
Updated Vault IRSA guide to include working instructions for basic setup.

# Context
When integrating with Vault users were getting 403 errors after the token lease had expired

## Testing steps
Follow the steps in https://docs.solo.io/gloo-edge/main/guides/integrations/vault/aws-irsa-auth/ using a post v1.15.3 (probably easiest to use the latest beta) release to see the 403 errors.

Repeat with the code in this branch and with the watcher increment set in settings (this shows only the new value to add, not the complete YAML block):
```
  secretOptions:
    sources:
    - vault:
        aws:
          leaseInterval: 300
```

and the errors will not occur


# Checklist:
